### DISCUSSION TITLE
Pull main docs Content from 3rd-party API (a headless CMS)

### DISCUSSION THREAD

**Discussion by brob** [withastro/starlight#1790] [Q&A] [answered]
I've seen Ghost and Obsidian plugins, but they look like they add additional content (or override the sidebar specifically).

Is there anyway to make main aspects of Starlight pull data from a third party API (headless CMS). I'd love to work on a plugin to integrate with mine, but I want to make sure it's possible in a meaningful way before I dive in.

I feel like a lot of the basic defaults won't work properly, but I could be wrong

**Comment by HiDeoo** [comment, marked as answer]
If I understand this correctly, you may be looking for [custom pages](https://starlight.astro.build/guides/pages/#custom-pages) using the [`<StarlightPage />`](https://starlight.astro.build/guides/pages/#using-starlights-design-in-custom-pages) component to use starlight’s design and feature in them. You will be able to rely on [dynamic routes](https://docs.astro.build/en/guides/routing/#static-ssg-mode) to generate pages as you need based on any data source.

This is a bit of a high-level overview, but I hope it helps and see if it’s at least a step in the right direction for you. Would be happy to help further if you have more specific questions or need more guidance or if I just misunderstood your question.

**Reply by brob** [reply]
The more I think about it, the more I think I can do it with custom pages potentially. I'd originally been thinking about wanting someone to be able to install a plugin and give a specific query to the CMS to get their pages and then have starlight do its thing, but I may have been overthinking it

**Reply by HiDeoo** [reply]
Not to say this a better approach or not, but more to give an example for comparison, this is also something doable. This is kinda what is done for example in the `starlight-openapi` plugin.

1. The plugin [adds an Astro integration](https://github.com/HiDeoo/starlight-openapi/blob/main/packages/starlight-openapi/index.ts#L22)
1. The integration [injects a dynamic route](https://github.com/HiDeoo/starlight-openapi/blob/main/packages/starlight-openapi/libs/integration.ts#L11-L15)
1. The integration [adds a Vite plugin](https://github.com/HiDeoo/starlight-openapi/blob/main/packages/starlight-openapi/libs/integration.ts#L17-L21) and pass down to it some data (could be configuration from the user at the plugin level, data from any source, etc.)
1. The Vite plugin [exposes such data](https://github.com/HiDeoo/starlight-openapi/blob/main/packages/starlight-openapi/libs/vite.ts#L5-L21) throught a virtual module
1. The injected route [defines the list of dynamic paths to generate](https://github.com/HiDeoo/starlight-openapi/blob/main/packages/starlight-openapi/components/Route.astro#L14-L16)
1. The injected route [renders Starlight custom pages](https://github.com/HiDeoo/starlight-openapi/blob/main/packages/starlight-openapi/components/Route.astro#L29-L31), it can reads the data from the virtual module, fetch more data, w/e is needed to render the page

**Comment by sgalcheung** [comment]
Also have this idea, how can we integrate a CMS/headless CMS, the **starlight-openapi** plugin seems not friendly with CMS, in one starlight, each directory corresponds to an article and requires an HTTP request, and we may have more than one column, corresponding to more than one startlight, and we will have many of these.

**Comment by sgalcheung** [comment]
Here is a project([Docuna](https://docuna.pages.dev/getting-started/headlesscms/)) also mention **integration your astro starlight project with headless cms**

**Comment by sgalcheung** [comment]
I start an experiment for this,  [starlight-squidex](https://github.com/sgalcheung/starlight-squidex)

Here are some puzzles for me:
- I don't know how I can inject data from a work project, so I added squidex to fetch data
- when building the [example](https://github.com/sgalcheung/starlight-squidex/tree/main/example), I see an error, I don't use the collection data, I render it from Squidex or any API you can.
```
[vite]: Rollup failed to resolve import "/Users/sgalcheung/***/starlight-squidex/example/src/content/config.ts" from "virtual:starlight/collection-config".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
```

**Comment by sgalcheung** [comment]
Just pass in the parameter sidebar saved me, like this
`<StarlightPage frontmatter={frontmatter} sidebar={sidebar}>`

And, I use Redis to cache dynamic load sidebar data.
